fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java

```diff
@@ -0,0 +1,411 @@
+                } else {
+                    insertDecodeExpr(exchangeExpr, Collections.singletonList(newChildExpr), 0, context);
+                }
+            }
+            return exchangeExpr;
+        }
+    }
+
+    @Override
+    public OptExpression rewrite(OptExpression root, TaskContext taskContext) {
+        if (!ConnectContext.get().getSessionVariable().isEnableLowCardinalityOptimize()) {
+            return root;
+        }
+        List<LogicalOlapScanOperator> scanOperators = taskContext.getAllScanOperators();
+
+        for (LogicalOlapScanOperator scanOperator : scanOperators) {
+            OlapTable table = (OlapTable) scanOperator.getTable();
+            long version = table.getPartitions().stream().map(Partition::getVisibleVersion)
+                    .max(Long::compareTo).orElse(0L);
+            List<ColumnRefOperator> outputColumns = scanOperator.getOutputColumns();
+            for (ColumnRefOperator column : outputColumns) {
+                // Condition 1:
+                if (!column.getType().isVarchar()) {
+                    continue;
+                }
+
+                ColumnStatistic columnStatistic = Catalog.getCurrentStatisticStorage().
+                        getColumnStatistic(table, column.getName());
+                // Condition 2:
+                if (!FeConstants.USE_MOCK_DICT_MANAGER && (columnStatistic.isUnknown() ||
+                        columnStatistic.getDistinctValuesCount() > CacheDictManager.LOW_CARDINALITY_THRESHOLD)) {
+                    LOG.debug("{} isn't low cardinality string column", column.getName());
+                    continue;
+                }
+
+                // Condition 3:
+                if (IDictManager.getInstance().hasGlobalDict(table.getId(), column.getName(), version)) {
+                    if (!tableIdToStringColumnIds.containsKey(table.getId())) {
+                        List<Integer> integers = Lists.newArrayList();
+                        integers.add(column.getId());
+                        tableIdToStringColumnIds.put(table.getId(), integers);
+                    } else {
+                        tableIdToStringColumnIds.get(table.getId()).add(column.getId());
+                    }
+                } else {
+                    LOG.debug("{} doesn't have global dict", column.getName());
+                }
+            }
+        }
+
+        if (tableIdToStringColumnIds.isEmpty()) {
+            return root;
+        }
+
+        DecodeContext context = new DecodeContext(tableIdToStringColumnIds, taskContext.getOptimizerContext().
+                getColumnRefFactory());
+
+        OptExpression rewriteExpr = root.getOp().accept(new DecodeVisitor(), root, context);
+        if (context.hasChanged) {
+            return generateDecodeOExpr(context, Collections.singletonList(rewriteExpr));
+        }
+        return rewriteExpr;
+    }
+
+    public static void insertDecodeExpr(OptExpression parentExpr, List<OptExpression> childExpr,
```

Comment:
Decode is a operator, use `insertDecodeOperator` is better?

Reply:
The finial physical tree is a OptExpression tree, not an operator tree. Here insert is a OptExpression, not an operator.